connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala

```diff
@@ -189,94 +158,164 @@ object KustoWriter{
+                                                    kustoAdminClient: Client,
+                                                    tmpTableName: String,
+                                                    isAsync: Boolean = false): Unit = {
+    import coordinates._
     try {
       // Move data to real table
       // Protect tmp table from merge/rebuild and move data to the table requested by customer. This operation is atomic.
       kustoAdminClient.execute(database, generateTableAlterMergePolicyCommand(tmpTableName, allowMerge = false, allowRebuild = false))
-      kustoAdminClient.execute(database, generateTableMoveExtentsCommand(tmpTableName, table))
+      kustoAdminClient.execute(database, generateTableMoveExtentsCommand(tmpTableName, table.get))
       KDSU.logInfo(myName, s"write to Kusto table '$table' finished successfully $batchIdIfExists")
     }
     catch {
       case exception: Exception =>
-        KDSU.reportExceptionAndThrow(myName, exception, "finalizing write operation", cluster, database, table, isLogDontThrow = isAsync)    }
+        KDSU.reportExceptionAndThrow(myName, exception, "finalizing write operation", cluster, database, table.get, isLogDontThrow = isAsync)
+    }
     finally {
       tryCleanupIngestionByproducts(database, kustoAdminClient, tmpTableName)
     }
   }
 
   private def ingestToTemporaryTableByWorkers(
-    batchId: Option[Long],
-    cluster: String,
-    database: String,
-    table: String,
-    schema: StructType,
-    batchIdIfExists: String,
-    timeout: FiniteDuration,
-    ingestRowsIntoKusto: (StructType, Iterator[InternalRow]) => IngestionResult, rows: Iterator[InternalRow]) = {
-
+                                               batchId: String,
+                                               timeout: FiniteDuration,
+                                               storageUri: String,
+                                               rows: Iterator[InternalRow])
+                                             (implicit parameters: KustoWriteResource)
+  : Unit = {
+
+    import parameters._
     val partitionId = TaskContext.getPartitionId
     KDSU.logInfo(myName, s"Ingesting partition '$partitionId'")
 
-      val ackIngestion: Future[IngestionResult] = Future {
-        ingestRowsIntoKusto(schema, rows)
-      }
+    // We force blocking here, since the driver can only complete the ingestion process
+    // once all partitions are ingested into the temporary table
+
+
+    Await.result(
+      Future(ingestRowsIntoKusto(rows, storageUri)).map { ingestionResults =>
+        // Proceed only on success. Will throw on failure for the driver to handle
+        ingestionResults.foreach { ingestionResult =>
+          KDSU.runSequentially[IngestionStatus](
+            func = () => ingestionResult.GetIngestionStatusCollection().get(0),
+            0, delayPeriodBetweenCalls, (timeout.toMillis / delayPeriodBetweenCalls + 5).toInt,
+            res => res.status == OperationStatus.Pending,
```

Comment:
How about the idea that we discussed to only pass result operation-ids to be aggregated by the driver, and the driver would test all results to completion (the driver also knows how many results to expect and can throw if the expected number is not matched within a few minutes)?

Reply:
Definetly. I will do it later